oauthenticator/globus.py

```diff
@@ -0,0 +1,123 @@
+"""
+Custom Authenticator to use Globus OAuth2 with JupyterHub
+"""
+import pwd
+from tornado import gen, web
+from tornado.auth import OAuth2Mixin
+from tornado.concurrent import return_future
+from tornado.web import HTTPError
+from .oauth2 import OAuthLoginHandler, OAuthenticator
+
+try:
+    import globus_sdk
+except:
+    raise HTTPError(500, ("Trying to use the Globus Auth "
+                          "authenticator, but globus_sdk "
+                          "is not installed"))
+
+
+class GlobusMixin(OAuth2Mixin):
+    """
+    Overriding the tornado function because
+    globus provides it's own method to assemble the
+    authorize url
+    """
+    @return_future
+    def authorize_redirect(self, client=None, callback=None):
+        self.redirect(client.oauth2_get_authorize_url())
+        callback()
+
+
+class GlobusLoginHandler(OAuthLoginHandler, GlobusMixin):
+    def get(self):
+        # Doing the scope weirdness for the globus_sdk
+        scopes = ['openid', 'profile', 'email',
+                  'urn:globus:auth:scope:auth.globus.org:view_identities']
+        scope_string = (' ').join(scopes)
+        redirect_uri = self.authenticator.get_callback_url(self)
+        client = self.authenticator.globus_portal_client()
+        client.oauth2_start_flow(
+            redirect_uri,
+            requested_scopes=scope_string,
+            refresh_tokens=True)
+        self.log.info('globus redirect: %r', redirect_uri)
+        self.authorize_redirect(client)
+
+
+class GlobusOAuthenticator(OAuthenticator):
+
+    login_service = "Globus"
+    login_handler = GlobusLoginHandler
+
+    def get_globus_app_id(self):
+        """
+        Getting the custom configuration value globus_app_id.
+        globus_app_id is the UUID of the Globus app to authenticate
+        against
+        """
+        return self.config.GlobusOAuthenticator.globus_app_id
```

Comment:
Looking at others, could I just rename it as `client_id` and `client_secret`? The mediawiki and  bitbucket example use those